File: api/routes/k8s-export.js
const express = require('express');
const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');
const fsp = fs.promises;

const router = express.Router();

const snapshotDir = path.join(__dirname, '../..', 'k8s-snapshot');

let exportState = {
  running: false,
  paused: false,
  pid: null,
  startedAt: null,
  totalNamespaces: 0,
  completedNamespaces: 0,
  activeNamespaces: new Set(),
  activeResources: new Set(),
  fileCount: 0,
  minEtaSeconds: null,
  error: null,
  output: '',
};

async function countFiles(dir) {
  let count = 0;
  try {
    const entries = await fsp.readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.name.startsWith('.') || entry.name.endsWith('.tmp')) continue;
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        count += await countFiles(full);
      } else {
        count++;
      }
    }
  } catch {
    // directory doesn't exist yet
  }
  return count;
}

async function countDoneNamespaces() {
  try {
    const entries = await fsp.readdir(snapshotDir, { withFileTypes: true });
    let count = 0;
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      try {
        await fsp.access(path.join(snapshotDir, entry.name, '.done'));
        count++;
      } catch { /* no .done marker */ }
    }
    return count;
  } catch {
    return 0;
  }
}

// POST /api/k8s-export/start
router.post('/k8s-export/start', (req, res) => {
  if (exportState.running) {
    return res.status(409).json({ error: 'Export already running' });
  }

  const resume = req.body?.resume === true;

  exportState = {
    running: true,
    paused: false,
    pid: null,
    startedAt: Date.now(),
    totalNamespaces: 0,
    completedNamespaces: 0,
    activeNamespaces: new Set(),
    activeResources: new Set(),
    fileCount: 0,
    error: null,
    output: '',
  };

  const useGo = process.env.USE_GO_EXPORT === 'true';
  const scriptPath = useGo
    ? path.join(__dirname, '../..', 'cmd', 'k8s-export', 'k8s-export')
    : path.join(__dirname, '../..', 'scripts', 'k8s-export.sh');

  let spawnCmd, args;
  if (useGo) {
    spawnCmd = scriptPath;
    args = [];
  } else {
    spawnCmd = 'bash';
    args = [scriptPath];
  }
  if (resume) args.push('--resume');

  const child = spawn(spawnCmd, args, {
    cwd: path.join(__dirname, '../..'),
    env: { ...process.env },
    detached: true,  // new process group — enables group kill on pause
  });

  exportState.pid = child.pid;

  child.stdout.on('data', (data) => {
    const raw = data.toString();
    process.stdout.write(raw);
    const text = raw.replace(/\x1b\[[0-9;]*m/g, '');
    // Cap output buffer to prevent memory bloat on long exports
    if (exportState.output.length < 200000) {
      exportState.output += text;
    }

    // Parse "Discovered N namespaces"
    const discoveredMatch = text.match(/Discovered (\d+) namespaces/);
    if (discoveredMatch) {
      exportState.totalNamespaces = parseInt(discoveredMatch[1], 10);
    }

    // Parse "=== Namespace: xxx === (complete, skipping)" — already done
    const skipMatches = text.matchAll(/=== Namespace: (.+?) === \(complete, skipping\)/g);
    const skippedSet = new Set();
    for (const m of skipMatches) {
      exportState.completedNamespaces++;
      skippedSet.add(m[1]);
    }

    // Parse "=== Namespace: xxx ===" — active export (not skipped)
    const nsMatches = text.matchAll(/=== Namespace: (\S+?) ===/g);
    for (const m of nsMatches) {
      if (!skippedSet.has(m[1])) {
        exportState.activeNamespaces.add(m[1]);
      }
    }

    // Parse "✓ Namespace xxx completed" — remove from active set
    const doneNsMatches = text.matchAll(/✓ Namespace (\S+) completed/g);
    for (const m of doneNsMatches) {
      exportState.activeNamespaces.delete(m[1]);
    }

    // Parse "→ fetching xxx" — resource/batch started downloading
    const fetchMatches = text.matchAll(/→ fetching (\S+)/gm);
    for (const m of fetchMatches) {
      for (const r of m[1].split(',')) {
        exportState.activeResources.add(r);
      }
    }

    // Parse "← xxx done" / "← xxx failed" — remove from active
    const doneResMatches = text.matchAll(/← (\S+) (?:done|failed)/gm);
    for (const m of doneResMatches) {
      for (const r of m[1].split(',')) {
        exportState.activeResources.delete(r);
      }
    }
  });

  child.stderr.on('data', (data) => {
    exportState.output += data.toString();
  });

  child.on('close', (code) => {
    exportState.running = false;
    exportState.pid = null;
    countFiles(snapshotDir).then(c => { exportState.fileCount = c; });
    if (code !== 0 && !exportState.paused) {
      exportState.error = `Process exited with code ${code}`;
    }
  });

  child.on('error', (err) => {
    exportState.running = false;
    exportState.pid = null;
    exportState.error = err.message;
  });

  res.json({ started: true, pid: child.pid, resume });
});

// GET /api/k8s-export/progress
router.get('/k8s-export/progress', async (req, res) => {
  const [liveCount, doneNs, hasCompleteMarker] = await Promise.all([
    countFiles(snapshotDir),
    countDoneNamespaces(),
    fsp.access(path.join(snapshotDir, '.export-complete')).then(() => true).catch(() => false),
  ]);

  // Derive totalNamespaces: prefer in-memory (from stdout parsing), fallback to filesystem
  let totalNamespaces = exportState.totalNamespaces;
  if (!totalNamespaces && doneNs > 0) {
    // Server restarted or stdout wasn't parsed — count namespace dirs as total
    try {
      const entries = await fsp.readdir(snapshotDir, { withFileTypes: true });
      totalNamespaces = entries.filter(e => e.isDirectory() && !e.name.startsWith('.')).length;
    } catch { /* ignore */ }
  }

  // Determine paused vs done when not running
  let paused = exportState.paused;
  if (!exportState.running && !exportState.paused) {
    if (hasCompleteMarker) {
      // .export-complete exists → export finished successfully
      paused = false;
    } else if (liveCount > 0) {
      // Files exist but no completion marker → partial/interrupted
      paused = true;
    }
  }

  // ETA: elapsed / doneNs * remainingNs — clamped to only decrease
  let etaSeconds = null;
  if (exportState.running && exportState.startedAt && doneNs > 0 && totalNamespaces > 0) {
    const elapsed = (Date.now() - exportState.startedAt) / 1000;
    const avgPerNs = elapsed / doneNs;
    const remaining = totalNamespaces - doneNs;
    const rawEta = Math.round(avgPerNs * remaining);
    // Clamp: ETA can only decrease, never jump upward
    if (exportState.minEtaSeconds === null || rawEta < exportState.minEtaSeconds) {
      exportState.minEtaSeconds = rawEta;
    }
    etaSeconds = exportState.minEtaSeconds;
  }

  const activeNsList = [...exportState.activeNamespaces];
  const response = {
    running: exportState.running,
    paused,
    totalNamespaces,
    completedNamespaces: doneNs,
    currentNamespace: activeNsList.join(', '),
    activeResources: [...exportState.activeResources],
    fileCount: liveCount,
    etaSeconds,
    error: exportState.error,
  };

  res.json(response);
});

// POST /api/k8s-export/stop — used for both pause and hard stop
router.post('/k8s-export/stop', (req, res) => {
  if (!exportState.running || !exportState.pid) {
    return res.status(400).json({ error: 'No export running' });
  }

  try {
    process.kill(-exportState.pid, 'SIGTERM');  // negative = kill entire process group
    exportState.running = false;
    exportState.paused = true;
    res.json({ stopped: true });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: cmd/server/routes/k8s_export.go
package routes

import (
	"encoding/json"
	"fmt"
	"io"
	"io/fs"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"syscall"
	"time"
)

// exportState tracks the running export process and its progress.
type exportState struct {
	mu                 sync.Mutex
	running            bool
	paused             bool
	pid                int
	startedAt          time.Time
	totalNamespaces    int
	completedNamespaces int
	activeNamespaces   map[string]struct{}
	activeResources    map[string]struct{}
	fileCount          int
	minEtaSeconds      *int
	err                string
	output             string
}

var state = &exportState{
	activeNamespaces: make(map[string]struct{}),
	activeResources:  make(map[string]struct{}),
}

var snapshotDir = "k8s-snapshot"

var (
	reDiscovered = regexp.MustCompile(`Discovered (\d+) namespaces`)
	reSkip       = regexp.MustCompile(`=== Namespace: (.+?) === \(complete, skipping\)`)
	reNs         = regexp.MustCompile(`=== Namespace: (\S+?) ===`)
	reDoneNs     = regexp.MustCompile(`✓ Namespace (\S+) completed`)
	reFetch      = regexp.MustCompile(`→ fetching (\S+)`)
	reDoneRes    = regexp.MustCompile(`← (\S+) (?:done|failed)`)
)

func registerK8sExport(mux *http.ServeMux) {
	mux.HandleFunc("POST /api/k8s-export/start", handleExportStart)
	mux.HandleFunc("GET /api/k8s-export/progress", handleExportProgress)
	mux.HandleFunc("POST /api/k8s-export/stop", handleExportStop)
}

// POST /api/k8s-export/start
// Spawns cmd/k8s-export binary (USE_GO_EXPORT=true) or scripts/k8s-export.sh.
// Parses stdout to track namespace progress and ETA.
func handleExportStart(w http.ResponseWriter, r *http.Request) {
	state.mu.Lock()
	if state.running {
		state.mu.Unlock()
		writeJSON(w, http.StatusConflict, map[string]any{"error": "Export already running"})
		return
	}

	var body struct {
		Resume bool `json:"resume"`
	}
	json.NewDecoder(r.Body).Decode(&body)

	state.running = true
	state.paused = false
	state.pid = 0
	state.startedAt = time.Now()
	state.totalNamespaces = 0
	state.completedNamespaces = 0
	state.activeNamespaces = make(map[string]struct{})
	state.activeResources = make(map[string]struct{})
	state.fileCount = 0
	state.minEtaSeconds = nil
	state.err = ""
	state.output = ""
	state.mu.Unlock()

	useGo := os.Getenv("USE_GO_EXPORT") == "true"
	var cmd *exec.Cmd
	if useGo {
		binary := filepath.Join("cmd", "k8s-export", "k8s-export")
		args := []string{}
		if body.Resume {
			args = append(args, "--resume")
		}
		cmd = exec.Command(binary, args...)
	} else {
		script := filepath.Join("scripts", "k8s-export.sh")
		args := []string{script}
		if body.Resume {
			args = append(args, "--resume")
		}
		cmd = exec.Command("bash", args...)
	}
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true} // new process group for group kill
	cmd.Dir = "."

	stdout, _ := cmd.StdoutPipe()
	stderr, _ := cmd.StderrPipe()

	if err := cmd.Start(); err != nil {
		state.mu.Lock()
		state.running = false
		state.err = err.Error()
		state.mu.Unlock()
		writeJSON(w, http.StatusInternalServerError, map[string]any{"error": err.Error()})
		return
	}

	state.mu.Lock()
	state.pid = cmd.Process.Pid
	state.mu.Unlock()

	go pipeOutput(stdout, false)
	go pipeOutput(stderr, true)
	go func() {
		cmd.Wait()
		count, _ := countFiles(snapshotDir)
		state.mu.Lock()
		state.running = false
		state.pid = 0
		state.fileCount = count
		if cmd.ProcessState != nil && cmd.ProcessState.ExitCode() != 0 && !state.paused {
			state.err = fmt.Sprintf("Process exited with code %d", cmd.ProcessState.ExitCode())
		}
		state.mu.Unlock()
	}()

	writeJSON(w, http.StatusOK, map[string]any{"started": true, "pid": cmd.Process.Pid, "resume": body.Resume})
}

// GET /api/k8s-export/progress
// Polled every 1s by frontend during export.
func handleExportProgress(w http.ResponseWriter, r *http.Request) {
	liveCount, _ := countFiles(snapshotDir)
	doneNs, _ := countDoneNamespaces(snapshotDir)

	_, errMarker := os.Stat(filepath.Join(snapshotDir, ".export-complete"))
	hasComplete := errMarker == nil

	state.mu.Lock()
	defer state.mu.Unlock()

	totalNs := state.totalNamespaces
	if totalNs == 0 && doneNs > 0 {
		// Server restarted — count namespace dirs as fallback.
		entries, err := os.ReadDir(snapshotDir)
		if err == nil {
			for _, e := range entries {
				if e.IsDir() && !strings.HasPrefix(e.Name(), ".") {
					totalNs++
				}
			}
		}
	}

	paused := state.paused
	if !state.running && !state.paused {
		if hasComplete {
			paused = false
		} else if liveCount > 0 {
			paused = true
		}
	}

	var etaSeconds *int
	if state.running && !state.startedAt.IsZero() && doneNs > 0 && totalNs > 0 {
		elapsed := time.Since(state.startedAt).Seconds()
		avgPerNs := elapsed / float64(doneNs)
		remaining := totalNs - doneNs
		raw := int(avgPerNs * float64(remaining))
		if state.minEtaSeconds == nil || raw < *state.minEtaSeconds {
			state.minEtaSeconds = &raw
		}
		etaSeconds = state.minEtaSeconds
	}

	activeNsList := make([]string, 0, len(state.activeNamespaces))
	for ns := range state.activeNamespaces {
		activeNsList = append(activeNsList, ns)
	}
	activeResList := make([]string, 0, len(state.activeResources))
	for res := range state.activeResources {
		activeResList = append(activeResList, res)
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"running":             state.running,
		"paused":              paused,
		"totalNamespaces":     totalNs,
		"completedNamespaces": doneNs,
		"currentNamespace":    strings.Join(activeNsList, ", "),
		"activeResources":     activeResList,
		"fileCount":           liveCount,
		"etaSeconds":          etaSeconds,
		"error":               state.err,
	})
}

// POST /api/k8s-export/stop
// Sends SIGTERM to the entire process group (negative PID).
func handleExportStop(w http.ResponseWriter, r *http.Request) {
	state.mu.Lock()
	defer state.mu.Unlock()

	if !state.running || state.pid == 0 {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "No export running"})
		return
	}

	if err := syscall.Kill(-state.pid, syscall.SIGTERM); err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]any{"error": err.Error()})
		return
	}

	state.running = false
	state.paused = true
	writeJSON(w, http.StatusOK, map[string]any{"stopped": true})
}

// pipeOutput reads from a pipe, writes to stdout, and parses progress markers.
func pipeOutput(r io.Reader, _ bool) {
	buf := make([]byte, 4096)
	for {
		n, err := r.Read(buf)
		if n > 0 {
			raw := string(buf[:n])
			os.Stdout.WriteString(raw)
			// Strip ANSI colour codes for parsing.
			text := stripANSI(raw)

			state.mu.Lock()
			if len(state.output) < 200000 {
				state.output += text
			}

			if m := reDiscovered.FindStringSubmatch(text); m != nil {
				fmt.Sscanf(m[1], "%d", &state.totalNamespaces)
			}

			skipped := map[string]bool{}
			for _, m := range reSkip.FindAllStringSubmatch(text, -1) {
				state.completedNamespaces++
				skipped[m[1]] = true
			}
			for _, m := range reNs.FindAllStringSubmatch(text, -1) {
				if !skipped[m[1]] {
					state.activeNamespaces[m[1]] = struct{}{}
				}
			}
			for _, m := range reDoneNs.FindAllStringSubmatch(text, -1) {
				delete(state.activeNamespaces, m[1])
			}
			for _, m := range reFetch.FindAllStringSubmatch(text, -1) {
				for _, res := range strings.Split(m[1], ",") {
					state.activeResources[res] = struct{}{}
				}
			}
			for _, m := range reDoneRes.FindAllStringSubmatch(text, -1) {
				for _, res := range strings.Split(m[1], ",") {
					delete(state.activeResources, res)
				}
			}
			state.mu.Unlock()
		}
		if err != nil {
			break
		}
	}
}

var ansiRe = regexp.MustCompile(`\x1b\[[0-9;]*m`)

func stripANSI(s string) string {
	return ansiRe.ReplaceAllString(s, "")
}

func countFiles(dir string) (int, error) {
	count := 0
	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		name := d.Name()
		if d.IsDir() {
			return nil
		}
		if strings.HasPrefix(name, ".") || strings.HasSuffix(name, ".tmp") {
			return nil
		}
		count++
		return nil
	})
	return count, err
}

func countDoneNamespaces(dir string) (int, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return 0, nil
	}
	count := 0
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		if _, err := os.Stat(filepath.Join(dir, e.Name(), ".done")); err == nil {
			count++
		}
	}
	return count, nil
}
